### ISSUE TITLE
Display errors and disappearing items

### ISSUE THREAD

**Issue by Aalwein** [jaquadro/StorageDrawers#541]
I'm noticing an issue with how drawers are displaying items and interacting with upgrades. 

1. This screenshot shows that I have three items in a 2x2 drawer in the tooltip. The drawer itself shows only two, while the "detailed" view of the drawer also shows only two. It is also showing the wrong number of items (the tooltip is correct, verified by pulling all the items and also looking at an attached RS grid).

2. You'll also notice the upgrades are offset to the left of the first box, yet that box is valid and there is an additional "hidden" upgrade box available on the far right, a total of seven upgrades. Was this an addition somewhere along the line (the last version I used was in Beyond 1.10.2) or is this an error? I can actually add 7 upgrades to the drawer, whether or not they are actually doing anything I can't tell. 

3. Lastly, despite the drawers being locked with a key, if I look at the detailed view of a drawer's contents or add another item to a multi-item box (for example, if a 2x2 only has 2 items stored and I add a third), the other "zeroed" items that should stay locked disappear.

Let me know if I can provide more detail. I could take a video of me interacting with the drawers and displaying this issues if you want.

![2017-07-17_20 05 17](https://user-images.githubusercontent.com/6945095/28295002-9d667b38-6b2b-11e7-80ed-402bc040e1b4.png)

**Comment by jaquadro** [comment]
There are 7 upgrade slots now.  It looks like you're using a texture pack, so it's a bit out of date in that regard.

**Comment by Aalwein** [comment]
I figured it was a change you added (holy crap, with 6 emerald upgrades and a void now... so many items!). I just updated the Lithos resource pack myself for Storage Drawers to fix it.

I noticed all this was happening when I was setting up automated systems - such as a mob farm and a RFTools quarry - feeding items into the drawers and extracting the ores and such to pulverizer/furnace automation and ingots back into the drawers. I set everything up last night to quarry and process then noticed the issues (when I took that screenshot). I logged in today to find most of my drawers all screwy now. My compression drawers holding ingots no longer have ingots - some have blocks, some have nuggets, but all the ingots that were being automatically fed into the drawers have disappeared. I can take a block out, pop it back in, but the drawer doesn't break it down to block/ingot/nugget as expected. Most of the 2x2s that were getting items from the mob farm have emptied. Any drawer that was holding ore (processed out into ingots but in a locked slot for buffering) is now blank despite the lock.

It's all very weird. I'm not sure when or what version this started happening because up until last night I was just sort of tossing stuff into a box and moving on to my next project. I'm using the same setup I used in 1.10.2 aside from using xnet instead of enderio pipes since enderio hasn't hit 1.12 yet. It worked flawlessly in 1.10.2. Also, when I disconnect the drawers from the system (I break the RS external storage link and the xnet-to-slave link), the behavior remains.

Just to show you - the image attached shows my storage exactly as it appeared (except the lead drawer, that's the drawer I was adding/removing to test this issue) after I logged back in today. The image in my OP is the state it was in just before I logged out. This is running on a private server on my own computer. You can see across the top I had various mob farm items, far left was the skull drawers, and the three center compression drawers (gold, iron, tin) and copper drawer to the right of the controller are all showing only blocks, no ingots or nuggets - those were the only ores I was collecting from the quarry at the time.

![2017-07-18_09 09 09](https://user-images.githubusercontent.com/6945095/28318973-93b892a0-6b9a-11e7-8a55-eb249621e05f.png)

**Comment by Lost-Ninja** [comment]
Also seeing this issue. I'm using the latest Mod version with 1.12... using forge 14.21.1.2415

**Comment by Brettflan** [comment]
I've had this happen in 1.12 with Forge 14.21.1.2415 as well. Initially I noticed some items not showing up visually on the drawers, but still being there according to The One Probe and the Refined Storage system that I have the drawers hooked up to. This was with both standard 1x1 drawers and with parts of compacting drawers like reported above, such as having the compacting drawer only show one or two of the available conversions for what's in the drawer.
I wasn't too bothered by that, since the items were still there, and the drawers would eventually visually update when something changed.

However, just a short bit ago I found that several drawers are now actually really empty, not just showing up as such. This is in a singleplayer game where I know for certain that many of those items (if not all) were not used for anything and should still be there. For example, one drawer which is now empty was filled with a few thousand dirt, and I haven't used any dirt lately.

So it seems that drawers frequently just visually glitch and don't display their contents properly on the face of the drawer, but sometimes they also actually really get their contents wiped out.

**Comment by Werral** [comment]
**Storage Drawers**: 1.12-5.2.9

**MinecraftForge Version**: 14.21.1.2420

**Minecraft Version**:  1.12

**Client or Server?**:
- [ x] Client
- [ ] Server

**What happens / happened?**:
Leaving the game or allowing the chucks to unload seems to cause this issue. it's not only a visual issue, the actual items disappear...sometimes.

Cheers!

**Comment by Brettflan** [comment]
Hmm. I'd stopped playing Minecraft for a while in hopes that this would be addressed, but there was an update to the pack I'm playing in recently (Beyond the Frontier) which brought it to Forge 2426, so I figured I'd give it a shot again in case a Forge update might fix it. Unfortunately, this problem persists. Every time that I save and quit within the same chunk as the storage drawers, when I load the save back up again, one or two more drawers have been emptied of their contents, and of course some other drawers will just not be properly displaying their contents. It does seem to be quite consistent when saving and loading.

Here's hoping a fix for this is found soon. I'd just started using Storage Drawers recently once I realized how useful the compacting drawers and drawer controller can really be when hooked up to a Refined Storage / Applied Energistics system, but it seems like I picked a pretty bad time to try it out.

EDIT:
I tested a bit further and it's not being as consistent as it was a bit ago. However, it does seem like the problem of randomly emptying drawers goes away if I remove my External Storage (from Refined Storage) from being connected to the Drawer Controller. So it might be from something like another device trying to access the inventory of a Drawer Controller before the bank of drawers has been fully initialized, or something like that ... it's a thought, anyway.
The problem of items being visually missing (but actually still being in the drawers) still happens every time, whether or not anything is connected to the Drawer Controller. It also manifests most consistently with the Compacting Drawers specifically, but also the others occasionally.

**Comment by Protheus** [comment]
Same issue as Werral. 192 stacks of stone, reloaded the game and all gone. They were in a default drawer.

**Comment by MysticAura** [comment]
This is happening when all drawers are in the same chunk as well, so it isn't a different chunk sync issue. Constantly losing drawer contents is not all that pleasant, and definitely a serious bug.  I look forward to seeing this addressed once the dev has the time.

MC: 1.12
Forge: 2443
Mod Version: 1.12.5.2.9

Steps to reproduce: Fill drawer with random amount of an item. Run around and forget about it a while. After a random amount of time the drawer will be empty.
@jaquadro  This is kind of more important than working on slimeball compat

**Comment by ghost** [comment]
MC: 1.12
Forge: 2443
Mod Version: 1.12.5.2.9

having the same issue. First time, the icon did not show on the drawer, but One Probe showed there was contents, but later the drawer was actually empty.
![2017-08-26_12 55 40](https://user-images.githubusercontent.com/22838247/29741677-edf31790-8a71-11e7-85c1-9befe8322b6f.png)
![2017-08-26_14 27 03](https://user-images.githubusercontent.com/22838247/29741678-f25cb03e-8a71-11e7-85a8-9a6cde9aa699.png)

**Comment by rjenkins2** [comment]
I'm also seeing this problem.

**Minecraft**: 1.12.1
**Storage Drawers**: 1.12.1-5.3.0
**Forge**: 2462
**Modpack**: Modern Skyblock 2 v2.3.8

Server: Yes

It's very frustrating to have a whole drawer of items suddenly disappear. This is pretty early on, so I have  just plain birch drawers with nothing special attached.

**Comment by jaquadro** [comment]
Are any of you having trouble using TOP Addons?

**Comment by Brettflan** [comment]
TOP Addons is indeed in the Beyond The New Frontier pack which I've been playing. It updated not long ago to the version of TOP Addons released yesterday, so I'm guessing it is probably fixed now. I'm not sure when I'll have time to test it to be sure, though.

**Comment by Lost-Ninja** [comment]
Had the issue with and without TOP Addons, had the issue without TOP altogether.

**Comment by jaquadro** [comment]
Has anyone been able to produce the visual glitch or item loss with just the mod alone?

For all the running around I've done on the ATC server, I've never experienced either problem there, nor had it reported.  And I haven't seen this in any dev testing.  So it would be nice to know if there could be a mod interaction at hand (it might still be SD's fault).  I mentioned TOP Addons because it was just discovered the other day that they were overwriting the stacks in drawers.

**Comment by ghost** [comment]
For me, the problem **seems** to have been resolved with the latest TOP addons. But the problem has been intermittent and had strange behaviour in that it mattered in which drawer the items were, and it seemed to have had a preference for Nether Quartz in my case. I'd like to keep an eye on it for a few days, just to make sure it was TOP Addons.

**Comment by Aalwein** [comment]
I've got all mods updated to the latest 1.12.1 releases as of this post (Forge, SD, and anything TOP related, along with NEI and such), and the items literally disappearing seems to be fixed. However, the visual issue of them disappearing from both the inventory (shift+rclick) view of the drawer and the face of the drawer is still happening. With items in the drawer, if I shift+rclick on the drawer the items disappear from both the inventory view and the face of the drawer. If I place an item in the drawer the face view returns and operates as expected until I shift+rclick on it again. It seems to only happen when there is an even stack amount (ie. 64 stack of cobble) and only when it's more than a single stack (I think). Actually, it seems to occur when you hit exactly 2 full stacks. It basically makes it impossible to remember which drawer is holding what item once the display blanks out, but at least the items are no longer being eaten by the void.

The drawer lock also seems to be working as expected again, as well. I'll play around some more and report if anything changes.

**Comment by Wynadorn** [comment]
The visual issue occurs when the game is trying to display 128 or more items in the detail menu (shift+right click). 127 eggs or 127 cobblestone for example will not trigger the bug but 128 will.

I have been able to reproduce this on a fresh client & server. I have not been able to reproduce this bug in a single-player game.

**Client:** 1.12.1-forge1.12.1-14.22.0.2467
**Client Mods:** StorageDrawers-1.12.1-5.3.0.jar & Chameleon-1.12-4.1.3.jar
**Server:** forge-1.12.1-14.22.0.2467-universal.jar
**Server Mods:** StorageDrawers-1.12.1-5.3.0.jar & Chameleon-1.12-4.1.3.jar

**Comment by jaquadro** [comment]
I've now reproduced this in a dev client/server combo, and the fix will be released tomorrow.

Two failings at play here, really.  One being an incomplete adaptation of the 1.12 refactorings, and the other being more fallout from Mojang's ItemStack overhaul.  Like last time, this weird limit is caused by Mojang writing the ItemStack count over the network as a java byte.  I need to encode a special tag onto items with the real stack count before sending them over the wire.

**Comment by jaquadro** [comment]
This is for the visual-only problem.

I think there's pretty good evidence actual item destruction was caused by the TOP bug (but possibly other mod interactions too).  Someone will need to demonstrate otherwise with just SD/Chameleon.

**Comment by jaquadro** [comment]
Fixed in 5.3.1.

**Comment by EPAVAIPOCRLH** [comment]
im having this issue rn , how do i fix it?

**Comment by thawizanator** [comment]
@EPAVAIPOCRLH not sure if this helps but this happen to me today. my wife was on her pc playing in our skyfactory 3 world, she opens it to LAN and we play together. when i loaded in all of our storage drawers were not show me anything in them. but on her screen she could see the items in each drawer. I tried reloading textures (F3 +T) and reloading chunks (F3 +A) that didn't seem to work so I had her try that too, didn't do anything. I ran far away from the storage area and tried reloading at this time she closed her entire client which kicked me out and reloaded hers so i did the same since i had to wait for her to get back in. Once it all loaded up and i got back in everything showed up. So what I figure that fixed the problem is: 

Move your character far enough away that the storage area chunks unload, then reload textures and chunks.
Close the client and reopen
if you are running a server stop and restart the server

this may fix the issue, it did for me
